support-modules/acquisition-events/src/main/scala/com/gu/support/acquisitions/ga/GoogleAnalyticsServiceImpl.scala

```diff
@@ -0,0 +1,255 @@
+          "cd26" -> acquisition.labels.exists(_.contains("gift-subscription")), // gift subscription
+          "cd27" -> productCheckout,
+          "cd30" -> acquisition.labels.exists(_.contains("corporate-subscription")), // corporate subscription
+
+          // Custom metrics
+          "cm10" -> getSuccessfulSubscriptionSignUpMetric(conversionCategory),
+
+          // Google Optimize Experiment Id
+          "xid" -> goExp.testNames,
+          "xvar" -> goExp.variantNames,
+
+          // The GA conversion event
+          "t" -> "event",
+          "ec" -> conversionCategory.name, //Event Category
+          "ea" -> productName, //Event Action
+          "el" -> camelCase(acquisition.paymentFrequency.value), //Event Label
+          "ev" -> acquisition.amount.map(_.toInt).getOrElse(0).toString, //Event Value is an Integer
+
+          // Enhanced Ecommerce tracking https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#enhancedecom
+          "ti" -> tid,
+          "tcc" -> acquisition.promoCode.getOrElse(""), // Transaction coupon.
+          "pa" -> "purchase", //This is a purchase
+          "pr1nm" -> productName, // Product Name
+          "pr1ca" -> conversionCategory.description, // Product category
+          "pr1pr" -> annualisedValue, // Product Price - we are tracking the annualised value here as this is what goes into the revenue metric
+          "pr1qt" -> "1", // Product Quantity
+          "pr1cc" -> acquisition.promoCode.getOrElse(""), // Product coupon code.
+          "pr1cm15" -> acquisition.amount.toString, // Custom metric 15 - purchasePrice
+          "cu" -> acquisition.currency.toString // Currency
+        )
+
+        body
+          .filter { case (key, value) => value != "" }
+          .map { case (key, value) => s"$key=$value" }
+          .mkString("&")
+    }
+  }
+
+  private[ga] def sanitiseClientId(maybeId: String): Either[BuildError, String] = {
+    maybeId match {
+      case clientIdPattern(id) => Right(id) // If we have a full _ga cookie string extract the client id
+      case "" => Left(BuildError("Client ID cannot be an empty string.\n" +
+        "To link server side with client side events you need to pass a valid clientId from the `_ga` cookie.\n" +
+        "Otherwise you can pass any non-empty string eg. `UUID.randomUUID().toString`\n" +
+        "More info here: https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#cid"))
+      case _ => Right(maybeId) // Otherwise assume that the caller has passed in the client id correctly
+    }
+  }
+
+  private[ga] def getProductCheckout(acquisition: AcquisitionDataRow): Option[String] =
+    acquisition.product match {
+      case Contribution | RecurringContribution => Some("Contribution")
+      case DigitalSubscription => Some("DigitalPack")
+      case Paper => getProductCheckoutForPrint(acquisition.printOptions)
+      case default => None
```

Comment:
I can actually see Guardian weekly was added below - so what other products are possible?

Reply:
Good question. I'm not sure I understand the existing logic.
The original thrift `Product` type had the following:
```
    ophan.thrift.event.Product.Contribution,
    ophan.thrift.event.Product.RecurringContribution,
    ophan.thrift.event.Product.MembershipSupporter,
    ophan.thrift.event.Product.MembershipPatron,
    ophan.thrift.event.Product.MembershipPartner,
    ophan.thrift.event.Product.DigitalSubscription,
    ophan.thrift.event.Product.PaperSubscriptionEveryday,
    ophan.thrift.event.Product.PaperSubscriptionSixday,
    ophan.thrift.event.Product.PaperSubscriptionWeekend,
    ophan.thrift.event.Product.PaperSubscriptionSunday,
    ophan.thrift.event.Product.PrintSubscription,
    ophan.thrift.event.Product.AppPremiumTier
```
So `GuardianWeekly` wasn't in this set.
With the new model, it is:
```
  case object Contribution extends AcquisitionProduct("CONTRIBUTION")
  case object RecurringContribution extends AcquisitionProduct("RECURRING_CONTRIBUTION")
  case object DigitalSubscription extends AcquisitionProduct("DIGITAL_SUBSCRIPTION")
  case object Paper extends AcquisitionProduct("PRINT_SUBSCRIPTION")
  case object GuardianWeekly extends AcquisitionProduct("PRINT_SUBSCRIPTION")
```

But `GuardianWeekly` is also a valid `PrintProduct`, so I'm not sure what the correct behaviour should be here.

Reply:
I've pushed up a possible solution: https://github.com/guardian/support-frontend/pull/3261/commits/83115e09958aa0ddb9dfd39fc4f86799e3d6fb0a

Reply:
that looks much better, however we are returning an Option, which breaks the types I think.  I will go and comment on the relevant line.